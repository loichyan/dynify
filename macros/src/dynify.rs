use proc_macro2::TokenStream;
use quote::{format_ident, quote, ToTokens};
use syn::{parse_quote_spanned, FnArg, Ident, Lifetime, Result, ReturnType, Token, Type};

use crate::lifetime::TraitContext;
use crate::utils::*;

pub fn expand(attr: TokenStream, input: TokenStream) -> Result<TokenStream> {
    let rename = syn::parse2::<Option<Ident>>(attr)?;
    let input_item = syn::parse2::<syn::Item>(input.clone())?;
    let output = match input_item {
        syn::Item::Trait(t) => expand_trait(rename, t)?,
        syn::Item::Fn(f) => expand_fn(rename, f)?,
        item => {
            return Err(syn::Error::new_spanned(
                &item,
                "expected a `fn` or `trait` item",
            ))
        },
    };
    Ok(quote!(#input #output))
}

fn expand_trait(rename: Option<Ident>, mut dyn_trait: syn::ItemTrait) -> Result<TokenStream> {
    let dyn_trait_name = rename.unwrap_or_else(|| format_ident!("Dyn{}", dyn_trait.ident));
    let input_trait_name = std::mem::replace(&mut dyn_trait.ident, dyn_trait_name);
    let dyn_trait_name = &dyn_trait.ident;

    let impl_target = format_ident!("{}Implementor", input_trait_name);
    let mut trait_impl_items = TokenStream::new();

    let (_, ty_generics, where_clause) = dyn_trait.generics.split_for_impl();
    for item in dyn_trait.items.iter_mut() {
        let impl_item = match item {
            syn::TraitItem::Const(syn::TraitItemConst {
                attrs,
                const_token,
                ident,
                colon_token,
                ty,
                semi_token,
                ..
            }) => {
                let attrs = attrs.outer();
                quote!(#(#attrs)* #const_token #ident #colon_token #ty
                    = #impl_target::#ident #semi_token)
            },
            syn::TraitItem::Type(syn::TraitItemType {
                attrs,
                type_token,
                ident,
                generics,
                semi_token,
                ..
            }) => {
                let attrs = attrs.outer();
                let (impl_generics, ty_generics, where_clause) = generics.split_for_impl();
                quote!(#(#attrs)* #type_token #ident #impl_generics
                    = #impl_target::#ident #ty_generics #where_clause #semi_token)
            },
            syn::TraitItem::Fn(syn::TraitItemFn { attrs, sig, .. }) => {
                let context = TraitContext {
                    generics: &dyn_trait.generics,
                };
                let transformed = transform_fn(Some(&context), sig, false)?;
                // TODO: support `#[dynify(skip)]`
                // TODO: support nested `#[dynify]`
                let attrs_outer = attrs.outer();
                let attrs_inner = attrs.inner();
                let target = quote_with(|tokens| {
                    impl_target.to_tokens(tokens);
                    NewToken![::].to_tokens(tokens);
                    sig.ident.to_tokens(tokens);
                });
                let impl_body = quote_transformed_body(transformed, &target, sig);
                quote!(#(#attrs_outer)* #sig { #(#attrs_inner)* #impl_body })
            },
            _ => continue,
        };
        trait_impl_items.extend(impl_item);
    }

    let impl_generics = quote_impl_generics(&dyn_trait.generics);
    Ok(quote!(
        #[allow(async_fn_in_trait)]
        #[allow(clippy::type_complexity)]
        #dyn_trait

        #[allow(clippy::type_complexity)]
        impl<#impl_generics #impl_target: #input_trait_name #ty_generics>
        #dyn_trait_name #ty_generics for #impl_target
        #where_clause { #trait_impl_items }
    ))
}

fn expand_fn(rename: Option<Ident>, mut dyn_fn: syn::ItemFn) -> Result<TokenStream> {
    let syn::ItemFn {
        vis,
        sig,
        attrs,
        block: _,
    } = &mut dyn_fn;

    let dyn_fn_name = rename.unwrap_or_else(|| format_ident!("dyn_{}", sig.ident));
    let input_fn_name = std::mem::replace(&mut sig.ident, dyn_fn_name);

    let transformed = transform_fn(None, sig, true)?;
    let attrs_outer = attrs.outer();
    let attrs_inner = attrs.inner();
    let impl_body = quote_transformed_body(transformed, &input_fn_name, sig);
    Ok(quote!(#(#attrs_outer)* #vis #sig { #(#attrs_inner)* #impl_body }))
}

/// Generates implementation body for a transformed function.
fn quote_transformed_body(
    transformed: TransformResult,
    target: &dyn ToTokens,
    sig: &syn::Signature,
) -> impl ToTokens {
    let arg_idents = sig.inputs.pairs().map(|p| {
        quote_with(move |tokens| {
            match p.value() {
                FnArg::Receiver(r) => r.self_token.to_tokens(tokens),
                FnArg::Typed(t) => t.pat.to_tokens(tokens),
            }
            p.punct_or_default().to_tokens(tokens);
        })
    });

    match transformed {
        TransformResult::Noop if sig.asyncness.is_some() => {
            quote!(#target (#(#arg_idents)*).await)
        },
        TransformResult::Noop => {
            quote!(#target (#(#arg_idents)*))
        },
        TransformResult::Function | TransformResult::Method => {
            let recv = sig.receiver().map(|r| &r.self_token);
            quote!(::dynify::__from_fn!([#recv] #target, #(#arg_idents)*))
        },
    }
}

/// Prints generics for implementation without angle brackets.
fn quote_impl_generics(generics: &syn::Generics) -> impl '_ + ToTokens {
    quote_with(move |tokens| {
        let is_lifetime = |p: &syn::GenericParam| matches!(p, syn::GenericParam::Lifetime(_));
        generics
            .params
            .pairs()
            .filter(|p| is_lifetime(p.value()))
            .chain(generics.params.pairs().filter(|p| !is_lifetime(p.value())))
            .for_each(|p| {
                p.value().to_tokens(tokens);
                p.punct_or_default().to_tokens(tokens);
            });
    })
}

#[derive(Clone, Copy)]
enum TransformResult {
    Noop,
    Function,
    Method,
}

/// Transforms the supplied function into a dynified one, returning `true` only
/// if the transformation is successful.
fn transform_fn(
    context: Option<&TraitContext>,
    sig: &mut syn::Signature,
    force: bool,
) -> Result<TransformResult> {
    let fn_span = sig.ident.span();
    if sig.asyncness.is_none() && get_impl_type(&sig.output).is_none() {
        if force {
            return Err(syn::Error::new(
                fn_span,
                "input function must return an `impl` type",
            ));
        } else {
            return Ok(TransformResult::Noop);
        }
    }

    let sealed_recv = match sig.receiver() {
        Some(r) => crate::receiver::infer_receiver(r)
            .ok_or_else(|| syn::Error::new(r.self_token.span, "unsupported receiver type"))
            .map(Some)?,
        None if force => None,
        None => return Ok(TransformResult::Noop),
    };

    let output_lifetime = Lifetime::new("'dynify", fn_span);
    crate::lifetime::inject_output_lifetime(context, sig, &output_lifetime)?;

    // Infer the appropriate output type
    let input_types = quote_with(|tokens| {
        sealed_recv
            .as_ref()
            .map(|r| quote!(::dynify::r#priv::#r,))
            .to_tokens(tokens);
        sig.inputs
            .pairs()
            .skip(sealed_recv.is_some() as usize)
            .for_each(|p| {
                match p.value() {
                    FnArg::Receiver(r) => r.ty.to_tokens(tokens),
                    FnArg::Typed(t) => t.ty.to_tokens(tokens),
                }
                p.punct_or_default().to_tokens(tokens);
            });
    });
    let output_type = match &sig.output {
        ReturnType::Default => ReturnType::Type(
            NewToken![->],
            parse_quote_spanned!(fn_span => ::dynify::r#priv::Fn<
                (#input_types),
                dyn #output_lifetime + ::core::future::Future<Output = ()>
            >),
        ),
        ReturnType::Type(r, ty) if sig.asyncness.is_some() => ReturnType::Type(
            *r,
            parse_quote_spanned!(fn_span => ::dynify::r#priv::Fn<
                (#input_types),
                dyn #output_lifetime + ::core::future::Future<Output = #ty>
            >),
        ),
        ty @ ReturnType::Type(..) => {
            let (r, ty) = get_impl_type(ty).unwrap();
            let bounds = ty
                .bounds
                .pairs()
                .filter(|p| !matches!(p.value(), syn::TypeParamBound::Lifetime(_)));
            ReturnType::Type(
                r,
                parse_quote_spanned!(fn_span => ::dynify::r#priv::Fn<
                    (#input_types),
                    dyn #output_lifetime + #(#bounds)*
                >),
            )
        },
    };

    sig.output = output_type;
    sig.asyncness = None;

    Ok(sealed_recv
        .map(|_| TransformResult::Method)
        .unwrap_or(TransformResult::Function))
}

fn get_impl_type(ty: &ReturnType) -> Option<(Token![->], &syn::TypeImplTrait)> {
    as_variant!(ty, ReturnType::Type(r, t))
        .and_then(|(r, ty)| as_variant!(&**ty, Type::ImplTrait).map(|ty| (*r, ty)))
}

#[cfg(test)]
#[path = "dynify_tests.rs"]
mod tests;
