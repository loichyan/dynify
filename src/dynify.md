A helper macro to create dyn compatible variants for traits.

`#[dynify]` accepts a trait or function as input and returns it as is, along
with a *dynified* variant. This involves converting each input function into a
function constructor. Consider the following example:

```rust
#[dynify::dynify]
trait Client {
    async fn request(&self, uri: &str) -> String;
}
```

It expands to something like this:

```rust
# use core::future::Future;
# use dynify::{Fn, from_fn};
trait Client {
    async fn request(&self, uri: &str) -> String;
}
// A dynified trait is generated and implemented for any type that implements
// `Client`. Although this can be done without this macro, it helps stay
// synchronized with the original trait, eliminating the need to handle elided
// lifetimes, which can be quite annoying for signatures with many lifetimes.
trait DynClient {
    fn request<'this, 'uri, 'dynify>(
        &'this self,
        uri: &'uri str,
    ) -> Fn!(&'this Self, &'uri str => dyn 'dynify + Future<Output = String>)
    where
        'this: 'dynify,
        'uri: 'dynify;
}
impl<ClientImplementor: Client> DynClient for ClientImplementor {
    fn request<'this, 'uri, 'dynify>(
        &'this self,
        uri: &'uri str,
    ) -> Fn!(&'this Self, &'uri str => dyn 'dynify + Future<Output = String>)
    where
        'this: 'dynify,
        'uri: 'dynify,
    {
        from_fn!(ClientImplementor::request, self, uri)
    }
}
```

## Customizing the generated traits

You can specify an identifier as the name of the generated trait:

```rust
#[dynify::dynify(MyDynClient)]
trait Client {
    async fn request(&self, uri: &str) -> String;
}
async fn run(client: &dyn MyDynClient) {
    // ...
}
```

The identifier must be supplied as the first argument.

## Lifetime conventions

The core feature of `#[dynify]` is the expansion of
[elided lifetimes](https://doc.rust-lang.org/nomicon/lifetime-elision.html).
`#[dynify]` employs a deterministic method for this:

1. The name of each elided lifetime is based on the name of the argument that
   contains it. Therefore, the binding pattern of each argument must be an
   identifier.
2. If a lifetime is the only one, it is named exactly the same as the argument.
3. Otherwise, an index is appended to each lifetime. The index is zero-based and
   corresponds to the order in which a lifetime occurs. Both elided and named
   lifetimes are taken into account when counting indices.
4. For a method receiver, the expanded lifetime is always named based on `this`
   instead of `self`. Rule *(3)* applies in this case as well.
5. If the return type is a dyn object, it always has a bound named `'dynify`.

The following demonstrates how these rules work:

```rust
# use dynify::Fn;
# use core::future::Future;
//#[dynify::dynify]
trait Trait {
    async fn method(&self, foo: &str, bar: (&str, &str)) -> String;
}
trait DynTrait {
    fn method<'this, 'foo, 'bar0, 'bar1, 'dynify>(
        &'this self,
        foo: &'foo str,
        bar: (&'bar0 str, &'bar1 str),
    ) -> Fn!(
        &'this Self,
        &'foo str,
        (&'bar0 str, &'bar1 str)
        => dyn 'dynify + Future<Output = String>)
    where
        'this: 'dynify,
        'foo : 'dynify,
        'bar0: 'dynify,
        'bar1: 'dynify,
        // Add extra bounds here
    ;
}
```

In common cases, you can rely on the lifetimes generated by `#[dynify]`, adding
extra bounds as needed.

## Making generated traits [`Send`]able

Unlike `async-trait`, this macro does not provide support for adding `Send`
bounds to returned `Future`s (or any other `impl Trait`s). However, as
illustrated below, you can combine `#[dynify]` with
[trait-variant](https://crates.io/crates/trait-variant) to achieve this:

```rust
# use dynify::PinDynify;
// You can also use `#(make(SendClient: Send))`. However, in this case, you can
// no longer specify a name in `#[dynify]` because `#[trait_variant::make]`
// will generate two traits, which leads to conflicting trait definitions.
#[trait_variant::make(Send)]
#[dynify::dynify] // must be put within the scope of `#[trait_variant::make]`
trait Client {
    async fn request(&self, uri: &str) -> String;
}
fn run_client(
    client: &(dyn DynClient + Sync),
) -> impl '_ + std::future::Future<Output = ()> + Send {
    async move {
        client.request("http://magic/request").pin_boxed().await;
    }
}
```

As an alternative, you can also [bitte](https://crates.io/crates/bitte) for this
purpose:

<!-- TODO: enable doctest after we move to edition 2024 -->

```rust,ignore
# use dynify::PinDynify;
#[bitte::bitte(Send)]
#[dynify::dynify] // must be put within the scope of `#[bitte]`
trait Client {
    async fn request(&self, uri: &str) -> String;
}
```
