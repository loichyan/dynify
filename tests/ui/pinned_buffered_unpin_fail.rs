use std::marker::PhantomPinned;
use std::ops::Deref;
use std::pin::Pin;

use dynify::{from_closure, Buffered, Dynify};

// Wraps Pin::into_inner to prevent rustc from reporting errors from rust-src.
fn unpin<P: Deref>(ptr: Pin<P>) -> P
where
    P::Target: Unpin,
{
    std::pin::Pin::into_inner(ptr)
}

fn main() {
    let mut stack = [0u8; 16];
    let init = from_closure(|slot| slot.write(PhantomPinned));
    let val: Buffered<PhantomPinned> = init.init(&mut stack);

    let pinned = std::pin::pin!(val);
    let _ = unpin(pinned); // fails
}
