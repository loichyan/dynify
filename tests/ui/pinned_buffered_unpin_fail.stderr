error[E0277]: `PhantomPinned` cannot be unpinned
  --> tests/ui/pinned_buffered_unpin_fail.rs:21:19
   |
21 |     let _ = unpin(pinned); // fails
   |             ----- ^^^^^^ the trait `Unpin` is not implemented for `Buffered<'_, PhantomPinned>`
   |             |
   |             required by a bound introduced by this call
   |
   = note: the trait bound `Buffered<'_, PhantomPinned>: Unpin` is not satisfied
   = note: required for `Buffered<'_, PhantomPinned>` to implement `Unpin`
note: required by a bound in `unpin`
  --> tests/ui/pinned_buffered_unpin_fail.rs:10:16
   |
8  | fn unpin<P: Deref>(ptr: Pin<P>) -> P
   |    ----- required by a bound in this function
9  | where
10 |     P::Target: Unpin,
   |                ^^^^^ required by this bound in `unpin`
help: consider borrowing here
   |
21 |     let _ = unpin(&pinned); // fails
   |                   +
21 |     let _ = unpin(&mut pinned); // fails
   |                   ++++
